### DISCUSSION TITLE
Conditional Trace Content/Message submission

### DISCUSSION THREAD

**Discussion by simonwh** [langfuse/langfuse#2578] [Ideas]
### Describe the feature or potential improvement

At Kin, we’re excited to integrate Langfuse for better observability of our new open-source agent. We are privacy-first and have a commitment to only share necessary user data - and be transparent about when it happens. That introduces a challenge with tracing:

**The Challenge**
Trace content (the messages) contain sensitive data, and we're committed to make it opt-in to share this data with us. The metrics around the traces itself can be logged without opt-in. We can control this today in the SDK by simply not sending the message content if user has not opted in.

However, if a user that has not opted in to share all data, does a "reports feedback" on a specific conversation/interaction, we want to ship the full trace (including message content) to Langfuse in order for us to understand what happened - and the use will give consent to share this as part of submitting the feedback.

**The Idea**
Introduce a way to conditionally ship contents/messages of traces if a specific action occurs, for example a user reports feedback, an error occurs or similar. 

Eager to hear your ideas/inputs on how this could potentially be implemented in Langfuse.

**Comment by marcklingen** [comment]
This is super interesting and a request I already heard a couple of times.

Core problems:
- Users report feedback in the frontend, but there is a delay between the backend tracing and the feedback. This delay can vary significantly.
- Storing the unsent events within the Langfuse SDK is likely insufficient, as each application instance would have its own queue of unsent events and would need to receive the "report feedback" information separately.

Potential solutions:
- Keep events in an "unsent queue" within the SDK and flush them to Langfuse when manually triggered by the "report feedback" action, based on the `traceId`. This approach does not work well if the backend service is deployed across multiple instances, as each instance would have its own queue of unsent events, and the "report feedback" signal would need to reach the same instance.
   - Additionally, this solution could lead to memory leakage, as the SDKs would need to store the unsent events for the expected period of "report feedback" signals (e.g., 10-20 minutes). After this period, no feedback could be reported, as the events would be deleted from the unsent queue.
- Send all traces to Langfuse, but only display and retain the data if there is a "report feedback" event within the first few minutes. This approach would ensure that all data hits the Langfuse service, but it may not be sufficient, and there could be a need to retain the data indefinitely, adding a `visible` flag to traces without "report feedback" events.
- The upcoming v3 of the system will introduce a Redis-based cache in the ingestion API. This cache could be used to store the events, and the data would be written to the database only if there is a manual "write to db" event within the first hour after the events were sent to the API. This approach would still send all data to the Langfuse service, but the data would not be persisted and could not be read unless the "write to db" event occurs.

Unfortunately given the complexity of getting this right, I cannot promise any short-term change which will enable this workflow. Is there an option for you to make a promise to your users to only _use_ the traces that they reported while being able to trace all executions?
